### PR TITLE
JDK-8325687: SimpleJavaFileObject specification would benefit from implSpec

### PR DESCRIPTION
**Pull request by jddarcy**
While reading the spec in the context of another potential change, noticed some cases where using implSpec would lead to a better structured specification.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8325904](https://bugs.openjdk.org/browse/JDK-8325904) to be approved

### Issues
 * [JDK-8325687](https://bugs.openjdk.org/browse/JDK-8325687): SimpleJavaFileObject specification would benefit from implSpec (**Enhancement** - P4)
 * [JDK-8325904](https://bugs.openjdk.org/browse/JDK-8325904): SimpleJavaFileObject specification would benefit from implSpec (**CSR**)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to [e5b57897](https://git.openjdk.org/jdk/pull/17854/files/e5b57897501613026d1a296406053351060a8488)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**) ⚠️ Review applies to [fe974f62](https://git.openjdk.org/jdk/pull/17854/files/fe974f620ec6b8d939e2102e5988c1a9ea1fb633)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17854/head:pull/17854` \
`$ git checkout pull/17854`

Update a local copy of the PR: \
`$ git checkout pull/17854` \
`$ git pull https://git.openjdk.org/jdk.git pull/17854/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17854`

View PR using the GUI difftool: \
`$ git pr show -t 17854`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17854.diff">https://git.openjdk.org/jdk/pull/17854.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17854#issuecomment-1944739905)

### CODE REVIEW

File: src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java [hunk 1]
```diff
@@ -79,31 +79,31 @@ public String getName() {
...
         throw new UnsupportedOperationException();
     }
 
     /**
-     * Wraps the result of {@linkplain #getCharContent} in a Reader.
-     * Subclasses can change this behavior as long as the contract of
-     * {@link FileObject} is obeyed.
+     * {@inheritDoc FileObject}
+     * @implSpec
+     * This implementation wraps the result of {@linkplain #getCharContent} in a Reader.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:106]: suggest `{@code Reader}` or `{@link Reader}`

File: src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java [hunk 2]
```diff
@@ -123,20 +123,20 @@ public Reader openReader(boolean ignoreEncodingErrors) throws IOException {
...
         throw new UnsupportedOperationException();
     }
 
     /**
-     * Wraps the result of openOutputStream in a Writer.  Subclasses
-     * can change this behavior as long as the contract of {@link
-     * FileObject} is obeyed.
+     * {@inheritDoc FileObject}
+     * @implSpec
+     * This implementation wraps the result of openOutputStream in a Writer.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:139]: Suggest using `{@code }` or `{@link }` for `openOutputStream` and `Writer`

File: src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java [hunk 3]
```diff
@@ -103,7 +103,8 @@ public OutputStream openOutputStream() throws IOException {
     /**
      * {@inheritDoc FileObject}
      * @implSpec
-     * This implementation wraps the result of {@linkplain #getCharContent} in a Reader.
+     * This implementation wraps the result of {@linkplain
+     * #getCharContent} in a {@link Reader}.
```
Comment (ExE-Boss) [inline src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:107]:
```suggestion
     * This implementation wraps the result of {@linkplain #getCharContent}
     * in a {@link Reader}.
```

File: src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java [hunk 4]
```diff
@@ -136,7 +137,8 @@ public CharSequence getCharContent(boolean ignoreEncodingErrors) throws IOExcept
     /**
      * {@inheritDoc FileObject}
      * @implSpec
-     * This implementation wraps the result of openOutputStream in a Writer.
+     * This implementation wraps the result of {@link
+     * openOutputStream} in a {@link Writer}.
```
Comment (ExE-Boss) [inline src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:141]:
```suggestion
     * This implementation wraps the result of {@linkplain #openOutputStream}
     * in a {@link Writer}.
```

**Review by lahodaj — APPROVED** [review]
Makes sense to me, thanks!

File: src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java (hunk 4 above)
Comment (ExE-Boss) [reply src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:141]: Still missing the `#` to make a valid link:
```suggestion
     * This implementation wraps the result of {@link #openOutputStream}
     * in a {@link Writer}.
```

Comment (jddarcy) [reply src/java.compiler/share/classes/javax/tools/SimpleJavaFileObject.java:141]: Hmm. Well, the build and docs build succeed with the `#` and produce a link in the generated HTML, but I'll add it before pushing. Thanks,

**Pull request closed without merging**